src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,323 @@
+                if (AnyTupleFieldNamesCollideWithExistingNames(
+                        semanticModel, tupleType, searchScope, cancellationToken))
+                {
+                    return false;
+                }
+
+                memberAccessExpressions = references.ToImmutable();
+                return true;
+            }
+            finally
+            {
+                references.Free();
+            }
+        }
+
+        private static bool AnyTupleFieldNamesCollideWithExistingNames(
+            SemanticModel semanticModel, INamedTypeSymbol tupleType,
+            SyntaxNode container, CancellationToken cancellationToken)
+        {
+            var existingSymbols = GetExistingSymbols(semanticModel, container, cancellationToken);
+
+            var reservedNames = semanticModel.LookupSymbols(container.SpanStart)
+                                             .Select(s => s.Name)
+                                             .Concat(existingSymbols.Select(s => s.Name))
+                                             .ToSet();
+
+            foreach (var element in tupleType.TupleElements)
+            {
+                if (reservedNames.Contains(element.Name))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static bool IsViableTupleTypeSyntax(TypeSyntax type)
+        {
+            if (type.IsVar)
+            {
+                // 'var t' can be converted to 'var (x, y, z)'
+                return true;
+            }
+
+            if (type.IsKind(SyntaxKind.TupleType))
+            {
+                // '(int x, int y) t' can be convered to '(int x, int y)'.  So all the elements
+                // need names.
+
+                var tupleType = (TupleTypeSyntax)type;
+                foreach (var element in tupleType.Elements)
+                {
+                    if (element.Identifier.IsKind(SyntaxKind.None))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool OnlyUsedToAccessTupleFields(
+            SemanticModel semanticModel, SyntaxNode searchScope, ILocalSymbol local,
+            ArrayBuilder<MemberAccessExpressionSyntax> memberAccessLocations, CancellationToken cancellationToken)
+        {
+            var localName = local.Name;
+
+            foreach (var identifierName in searchScope.DescendantNodes().OfType<IdentifierNameSyntax>())
+            {
+                if (identifierName.Identifier.ValueText == localName)
+                {
+                    var symbol = semanticModel.GetSymbolInfo(identifierName, cancellationToken).GetAnySymbol();
+                    if (local.Equals(symbol))
+                    {
+                        if (!(identifierName.Parent is MemberAccessExpressionSyntax memberAccess))
```

Comment:
What if we're inside a for loop that declares a local whose type name is the same as the tuple variable's name? Isn't the `Program` in the loop an IdentiferNameSYntax?
```
    class Program
    {
        static void Main(string[] args)
        {
            foreach ((int, int) Program in GetZ())
            {
                Program p = null;
                Console.WriteLine(p);
            }
        }

        static IEnumerable<(int, int)> GetZ() { }
    }
```

Reply:
(yes, very contrived)

Reply:
I'm not sure i understand your example.  The purpose of this feature is to see how the tuple-typed variable is used.  i.e. as long as you only use it to access tuple fields, then it's safe to elide the tuple altogether.  But, if you access things like methods on the tuple, or you use the variable in any other manner, then we assume it's not safe to use.

In your example I'm not sure what's going on (as you have a type and variable with teh same name).

Reply:
Oh, never mind--I can read.